lib/psci/psci_suspend.c

```diff
@@ -85,6 +85,14 @@ static void psci_suspend_to_pwrdown_start(unsigned int end_pwrlvl,
 	 */
 	cm_init_my_context(ep);
 
+	/*
+	 * Plat. management: Allow the platform to perform any early
+	 * actions required to power down the CPU, before we disable
+	 * the data cache.
+	 */
+	if (psci_plat_pm_ops->pwr_domain_suspend_early)
+		psci_plat_pm_ops->pwr_domain_suspend_early(state_info);
+
```

Comment:
OK, I agree the new hook has to go after `psci_spd_pm->svc_suspend()`. However, I'm not sure about your arguments for making this an optional hook on power down only.

For argument 1 above, platforms that require interconnect programming can continue to use `pwr_domain_suspend()`. Platforms that don't require this can use the new hook instead. Forcing the latter platforms to provide an empty `pwr_domain_suspend()` seems an over-restriction.

For argument 2 above, I think you mean to refer to `pwr_domain_suspend()` in the statement "therefore the `psci_spd_pm->svc_suspend()` needs to be used for pure retention suspend". Anyway, it would be possible to refactor the code to move `psci_do_pwrdown_sequence()` out of `psci_suspend_to_pwrdown_start()`, so that the hook is also called for retention states. But I agree the use-case for this is tenuous.

Argument 3 is purely subjective. I like the symmetry of a matching `pwr_domain_suspend_early()`, but I can see you don't. I think either way can work if it's documented properly.

I think this comes down to whether we think this new hook is all we're going to need for the foreseeable future or whether other matching hooks (e.g. for off and the finishers) are going to be needed. At the moment, I'm finding hard to see a need for other hooks and I can see your proposal would involve less code changes.

Does anyone else have an opinion?